{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "306\n",
      "195\n",
      "26028\n",
      "450\n",
      "191\n",
      "49\n",
      "66\n"
     ]
    }
   ],
   "source": [
    "def main():\n",
    "    virus = readGenome('lambda_virus.fa')\n",
    "    print(len(naive_with_rc('AGGT', virus)))\n",
    "    print(len(naive_with_rc('TTAA', virus)))\n",
    "    print(naive_with_rc('ACTAAGT', virus)[0])\n",
    "    print(naive_with_rc('AGTCGA', virus)[0])\n",
    "    print(len(naive_2mm('TTCAAGCC', virus)))\n",
    "    print(naive_2mm('AGGAGGTT', virus)[0])\n",
    "    sequences, qualities = readFastq('ERR037900_1.first1000.fastq')\n",
    "    print(lowest_quality_base(qualities))\n",
    "    \n",
    "\n",
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):  #Loop over alignments\n",
    "        match = True\n",
    "        for j in range(len(p)):  #Loop over characters\n",
    "            if t[i+j] != p[j]:  #Compare characters\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)  #All chars matched; record\n",
    "    return occurrences\n",
    "\n",
    "\n",
    "def reverseComplement(s):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N'}\n",
    "    t = ''\n",
    "    for base in s:\n",
    "        t = complement[base] + t\n",
    "    return t\n",
    "\n",
    "\n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            #Ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "\n",
    "\n",
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()  #Skip name line\n",
    "            seq = fh.readline().rstrip()  #Read base sequence\n",
    "            fh.readline()  #Skip placeholder line\n",
    "            qual = fh.readline().rstrip()  #Base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities\n",
    "\n",
    "\n",
    "def naive_with_rc(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):  #Loops over alignments\n",
    "        for seq in (p, reverseComplement(p)):\n",
    "            match = True\n",
    "            for j in range(len(seq)):  #Loops over characters\n",
    "                if t[i+j] != seq[j]:  #Compare characters\n",
    "                    match = False\n",
    "                    break\n",
    "            if match:\n",
    "                occurrences.append(i)  #All characters matched are added\n",
    "                break\n",
    "    return occurrences\n",
    "\n",
    "\n",
    "def naive_2mm(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):  #Loops over alignments\n",
    "        mismatches = 0\n",
    "        for j in range(len(p)):  #Loops over characters\n",
    "            if t[i+j] != p[j]:  #Compare characters\n",
    "                mismatches += 1\n",
    "                if mismatches > 2:\n",
    "                    break\n",
    "        if mismatches <= 2:\n",
    "            occurrences.append(i)  #Characters within two mismatches are added\n",
    "    return occurrences\n",
    "\n",
    "\n",
    "def lowest_quality_base(qs):\n",
    "    total = [0] * len(qs[0])\n",
    "    for q in qs:\n",
    "        for i, phred in enumerate(q):\n",
    "            total[i] += phred33ToQ(phred)\n",
    "    return total.index(min(total))\n",
    "\n",
    "\n",
    "def phred33ToQ(qual):\n",
    "    return ord(qual) - 33 #Measures phred quality of sequence\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
